test/test_jsonPath.js

```diff
@@ -42,71 +42,75 @@ request = require('supertest'),
+      this.checkBody('$.store.book[0].category', true).first().trim().eq('reference');
+      this.checkBody('$.store.book[*].price', true).filter(function(v,k,o){
+        return v > 10;
+      }).first().gt(10);
+
 			if(this.errors) {
-        // console.log(this.errors)
 				this.status=500;
 			}else{
 				this.status=200;
 			}
 		});
-		var req = request(app.listen());
+		var req = request(app.callback());
 		req.post('/json')
 		.send(data)
-		.expect(200 , done);
+		.expect(200 , cb);
+    function cb(stuff){
+      console.log(stuff)
+       done()
+    }
 	});
 
 });
 
 describe('koa-validate type' , function(){
   it("type check" , function(done){
-    var app = appFactory.create(1);
+    const app = appFactory.create(1);
     app.router.post('/json',function*(){
-      this.checkBody('/', true).notEmpty();
-      this.checkBody('/store/book[0]/price', true).get(0).type('number').type("primitive")
-      this.checkBody('/store/book[0]/price', true).get(0).type('hello') // should warn
-      this.checkBody('#/store/book[0]/category', true).first().type('string');
-      this.checkBody('/store/book[*]/price', true).type('array')
-      this.checkBody('/store/book[0]/publishDate', true).get(0).toDate().type('date').type('object')
+
+      this.checkBody('$.', true).notEmpty();
+      this.checkBody('$.store.book[0].price', true).get(0).type('number').type('primitive');
+      this.checkBody('$.store.book[0].price', true).get(0).type('hello'); // should warn
+      this.checkBody('$.store.book[0].category', true).first().type('string');
+      this.checkBody('$.store.book[*].price', true).type('array');
+      this.checkBody('$.store.book[0].publishDate', true).get(0).toDate().type('date').type('object');
+
       if(this.errors) {
         this.status=500;
       }else{
         this.status=200;
       }
     });
-    var req = request(app.listen());
+    var req = request(app.callback());
     req.post('/json')
     .send(data)
     .expect(200 , done);
   });
 it("type fail check" , function(done){
-    var app = appFactory.create(1);
+    const app = appFactory.create(1);
     app.router.post('/json',function*(){
-      this.checkBody('/', true).type('null');
-      this.checkBody('/store/book[0]/price', true).get(0).type('string');
-      this.checkBody('#/store/book[0]/category', true).first().type('null');
-      this.checkBody('/store/book[*]/price', true).type('nullorundefined')
-      this.checkBody('/store/book[0]/publishDate', true).first().toDate().type('array')
-      // console.log(this.errors)
+      this.checkBody('$.', true).type('null');
+      this.checkBody('$.store.book[0].price', true).get(0).type('string');
+      this.checkBody('$.store.book[0].category', true).first().type('null');
+      this.checkBody('$.store.book[*].price', true).type('nullorundefined');
+      this.checkBody('$.store.book[0].publishDate', true).first().toDate().type('array');
```

Comment:
This is the ONLY group of tests that are actually testing the functions where `transFn = true`, and none of them are testing the **checkQuery** function with it set to true.

This doesn't really have any impact on how we're using it, as we seem to always be calling these functions without that parameter...but if we're publishing it to the FOSS community, other users might be using it.

^^ This pertains to my questions/suggestions on the **getValue** and **hasKey** functions in validator.js. Those code changes you made are only accessed when `transFn` is true, but there are almost no tests around what you changed.

Reply:
What I mean is:
* you made changes to `getValue` and `hasKey` in **validate.js**
* Those two functions are only called from `checkBody` and `checkQuery`.
* _Neither of the two (and only two) tests that call_ `checkQuery` _(in **test_validate.js**) are testing with_ `transFn = true`.
* None of these tests are currently hitting the `hasKey` changes you made (I think because it still says `startsWith('/')`) -- so even if you change **validate.js** line 30 to `return 'MONKEY!!!!'` or `return undefined`, all of the existing tests still pass.

My first point was, the contract for `hasKey` is currently broken -- these tests aren't as comprehensive as they seemed, because they're not actually checking everything that you changed...but even if they _did_ traverse the path through `hasKey` correctly, the current tests obviously don't seem to care what that function returns.

My second point was, this doesn't really impact OUR usage of this code, because we never call `checkBody` or `checkQuery` with `transFn = true`.

My third (and final) point was, if we're going to publish this for the FOSS community, we might want to **at least** fix that `hasKey` function to match the same `startsWith` condition to match what you modified in `getValue`, or **at most** fix it and add some kind of unit tests around it to verify the contract isn't broken.